server/src/main/java/org/opensearch/discovery/PeerFinder.java

```diff
@@ -128,6 +139,28 @@ public PeerFinder(
         );
     }
 
+    public ActionListener<Void> nodeCommissionedListener() {
+        return new ActionListener<Void>() {
+            @Override
+            public void onResponse(Void unused) {
+                localNodeDecommissioned = false;
+                findPeersInterval = DISCOVERY_FIND_PEERS_INTERVAL_SETTING.get(settings);
+                logger.info("updated findPeersInterval to [{}] as node is commissioned", findPeersInterval);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                localNodeDecommissioned = true;
+                findPeersInterval = DISCOVERY_FIND_PEERS_INTERVAL_DURING_DECOMMISSION_SETTING.get(settings);
+                logger.info("updated findPeersInterval to [{}] as node is decommissioned", findPeersInterval);
+            }
+        };
+    }
+
+    public boolean localNodeDecommissioned() {
+        return localNodeDecommissioned;
+    }
+
```

Comment:
Seems like force-fitting a listener here. I would suggest you extend `CoordinatorPeerFinder` to `DecommissionedPeerFinder` and override intervals or properties as needed. Then in the `Coordinator` have an `AtomicReference` to `PeerFinder` and expose a private method in `Coordinator` as a `Consumer` to  `JoinHelper` to switch the PeerFinder from `CoordinatorPeerFinder` to `DecommissionedPeerFinder` on decommission exception

Reply:
Added a method in `PeerFinder` to update `findPeerIntervals` and other related field. Added a private method in `Coordinator` which is a `Consumer` to `JoinHelper`. Now, whenever Join fails with `NodeDecommissionedException`, the private method is invoked to call the method in `Peerfinder` which updates the `findPeerInterval`

Now, if I try to implement changes to have AtomicReference of PeerFinder in Coordinator it would break the contract at multiple places and lot of refactoring would be needed. Do you think it is absolutely necessary or we can use the Consumer to set the values as per need